### ISSUE TITLE
 'inst_type'  and 'inst_centroid'  

### ISSUE THREAD

**Issue by zerodohero** [vqdang/hover_net#64]
Sorry to bother you again, But I want to know how 'inst_type'  and 'inst_centroid' is generated in the tag?In particular, the center of mass, is there software that can automatically locate it?
**'inst_type' is a Nx1 array, indicating the type of each instance (in order of inst_map ID) 
 'inst_centroid' is a Nx2 array, giving the x and y coordinates of the centroids of each instance (in order of inst map ID).
there**

**Comment by zerodohero** [comment]
Also, is the 'inst_type' 'inst_centroid' in the tag used only for testing?I didn't see it during the training

**Comment by vqdang** [comment]
>how 'inst_type' and 'inst_centroid' is generated in the tag
https://github.com/vqdang/hover_net/blob/5d1560315a3de8e7d4c8122b97b1fe9b9513910b/src/process.py#L102

>'inst_type' 'inst_centroid' in the tag used only for testing
Yes it is

**Comment by zerodohero** [comment]
Thank you for your prompt reply
I would like to trouble you with another question. Now I have labeled some cells, such as the picture below. I have labeled all the same cells in the same color, and then I use OpenCV to convert them into matrix, that is, the "type_map" matrix in the program. It will have some sticky areas, as you've noted . Could you give me some directions if it is convenient for you?I will appreciate it and wish you all the best.

![image](https://user-images.githubusercontent.com/38395598/93410817-4ef64280-f8cc-11ea-8dcd-ab611de854ae.png)

**Comment by simongraham** [comment]
What do you mean by 'sticky areas' ? I assume that this is the label that you have generated and you also have the associated instance map where each object is uniquely labelled? Please give as much information as possible if you would like some guidance 👍 

For example:

- What dataset are you using
- How many images
- What image modality 
- What your aims are

Then we can give you some more clear directions.

**Comment by zerodohero** [comment]
Thank you for your reply.
The arrow in the picture indicates sticky areas.
![image](https://user-images.githubusercontent.com/38395598/93457345-6d326180-f911-11ea-9e6f-d438ea9a2652.png)
The data I used is the same as the data you used in your paper, which is the cell HE diagram, as shown below
![image](https://user-images.githubusercontent.com/38395598/93457560-b84c7480-f911-11ea-8bd0-bbf31de36001.png)
There are about 20 pieces of 1000 * 1000 data

My purpose is through the figure 1 mark location of the cell to generate matrix, namely "type_map", but the same color in the OpenCV can only generate the same number, which is the same kind of cells is the same color, it user the same number said, ut the "inst_map" matrix is represented by the unique number of each cell. How do you generate an "inst_map" matrix from  figure 1?

**Comment by zerodohero** [comment]
If each cell is manually labeled with a different color, this project is too large to be easily realized

**Comment by zerodohero** [comment]
Sorry to bother you again, but could you answer my question?

**Comment by simongraham** [comment]
Hi. I am still not completely clear on what you are trying to do here. Are you labelling this for the purpose of generating ground truth for training purposes? If so, then this needs to be done manually. We obtained an instance map by hand labelling data, which is a very time consuming and tedious process. Of course, when you hand mark the instance map, the cells will be separated because the user has control over this. For the type map, we use one colour per cell. So overall we have one inst_map that has a unique value per nucleus and a type_map where each nucleus has a unique value. For this, each pixel is also labelled. You can download the CoNSeP data from [this link](https://warwick.ac.uk/fac/sci/dcs/research/tia/data/hovernet/) and explore the data to see exactly what I mean from the comment above. 

Maybe I am misunderstanding your question, but please let me know if this helps. 

Also, it seems that the diagram that you showed above does not correspond to the image?

**Comment by zerodohero** [comment]
Thank you very much for your reply
yes, I labelling this for the purpose of generating ground truth for training purposes.  
Do you manually label the Numbers in the instance Map, as shown in the figure below (local matrix extracted from the instance map)?

![image](https://user-images.githubusercontent.com/38395598/93862740-1653d000-fcf5-11ea-8e63-04c95d0e86c4.png)
Instead of first drawing different color maps on the original HE diagram as shown below, and then generating an instance map matrix like the one above from this picture?
![image](https://user-images.githubusercontent.com/38395598/93862927-629f1000-fcf5-11ea-93de-679b5da63f3a.png)

**Comment by zerodohero** [comment]
In general, we want to generate the matrix by first drawing the color of the contour of the cell

**Comment by zerodohero** [comment]
However, the problem we encountered was that the color-generated matrix would have some corner errors, as shown below
![image](https://user-images.githubusercontent.com/38395598/93864190-16ed6600-fcf7-11ea-98c7-ce2d14803e4e.png)

**Comment by simongraham** [comment]
How are you generating the annotation? You can see that the colours are not homogeneous, which will lead to differences as you are highlighting above. There are multiple ways of annotating, but you need to ensure that pixels are labelled accurately.

**Comment by zerodohero** [comment]
According to the method you told me before, I made a mask on the cell, marked different cells with different colors, and then exported the mask, which was processed into the form of digital matrix by script.Here's what you said at the time.The first image above is the exported mask, but we marked the same cells in the same color first. 
**Hi @zerodohero

Thanks a lot for your kind words. We used the software GIMP to do the annotation and used the pencil tool to ensure there was no aliasing. For this, we opened the original image and annotated the nuclei with a transparency layer. We didn’t use a specific schema for choosing the colours for annotating, as long as touching nuclei were not annotated with the same colour. Then we coloured the non nuclei black and exported the transparency layer. Once this is done, you can use a script to create the nuclear instance map.

For generating the class map, we also used GIMP and used different colours for each class. Here, the nuclei outline does not need to be annotated as this has already done, but you can simply provide a dot with the appropriate colour**

Could you please provide me with more information? I can't figure out how to mark the H&E diagram with pixel-level Numbers directly. You suggested me to interpret it as what I said above, did I do something wrong?

**Comment by simongraham** [comment]
Yes, but you are doing something to make the colours not homogeneous within each nucleus. The eraser can introduce artefacts, which looks like you have used in this image. The errors that you are getting will be to do with how you have used GIMP. The map will need cleaning up.

**Comment by zerodohero** [comment]
Yes, we are also looking at how to make the color more uniform and remove as much interference as possible. Do you have any good Suggestions on this?We found that multiple application of the same color on the same cell (because one application may not be well done in detail) may exacerbate the color inhomogeneity.

**Comment by vqdang** [comment]
Regarding the eraser, set it to this will delete without feathering effect.
![image](https://user-images.githubusercontent.com/24943262/93890613-70976580-fce2-11ea-81e1-5774aa726100.png)

**Comment by vqdang** [comment]
Also, for brush, must use as follow to remove feathering effect.
![image](https://user-images.githubusercontent.com/24943262/93890768-97ee3280-fce2-11ea-8acf-25a9ac8211a9.png)

**Comment by zerodohero** [comment]
Thank you for sharing so unselfishly and wish you all the best
